Import authors (#93)
nickmoreton/wagtail-wordpress-import: docs/import_hooks.md

</item>
```

The type of an XML `<item>` element is given by its `<wp:post_type>` tag. The example item above represents an "attachment" type. This is not included in the import by default, because we only import items of the "post" or "page" type, representing a blog post or page to be created in Wagtail.

Import hooks allow you to define a function that will be called with some parameters where you can define your own custom import logic for your specific import requirements.

## Adding an import hook

The package doesn't include any import hooks by default.

> You can add your own import hooks by adding a settings to your own sites settings.

Reviewer: **Suggestion**: Reword

```suggestion
You can add your own import hooks by modifying:

- `settings.WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE` for tags within the `<wp:post_meta>` tag, or
- `settings.WORDPRESS_IMPORT_HOOKS_TAGS_TO_CACHE` for tags which are directly attributes of the imported post or page
```

## Example XML Item Hook Configuration

This hook is run for extra XML items that will be imported.

Add a settings variable to your site settings file called `WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE`, add a dict with a key of `attachment` and values as follows:

```python
WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE = {
    # the key is the wp:post_type XML tag value

>     "attachment": { 

Reviewer: **Question**: Is this the post_type of the target object to look up? It seems to say that on line 97 below. Can we make it clearer here?

        # the XML wp:postmeta tag wp:meta_key value
        "DATA_TAG": "thumbnail_id", 
        # the dotted path to the function to call in your own Wagtail site
        "FUNCTION": "pages.import_hooks.header_image_processor", 
    }
}
```

The key of the dict is the value in  `<wp:post_type>...</wp:post_type>` in the XML item that contains the related content.

`DATA_TAG` is the value in `<wp:meta_key>value</wp:meta_key>` of the `<wp:postmeta>` tag in the XML item that contains the related content. *This is passed to your function.*

`FUNCTION` the dotted path to the function to call in your own Wagtail site.

> Note: If the DATA_TAG has a leading `_` character do not include the underscore in the settings value. If the value contains a `:` anywhere in it replace it with a `_` in the settings value.

Reviewer: **Praise**: Ah, you mentioned this change. This is good.

**Suggestion**: I think this section might be clearer if we added the example four lines of XML before the example Python, and moved some of the descriptions to there. That would avoid lengthy references in prose, e.g.

> `DATA_TAG` is the value in `<wp:meta_key>value</wp:meta_key>` of the `<wp:postmeta>` tag in the XML item that contains the related content.

Author: **Question** did I miss this yesterday.

Sorry I'm not sure what you mean here.

Reviewer: I was suggesting rewording. GitHub wouldn't let me do a massive rewrite, but what's already here is fine.

### Sample function for processing the header image

```python
# pages/import_hooks.py

# get_or_save_image is a convenience function provided in the package
from wagtail_wordpress_import.block_builder_defaults import get_or_save_image

> 

Reviewer: **Suggestion**: PEP-8 two lines before a function definition
```suggestion

```

    <wp:post_name>a-title</wp:post_name>
    <wp:status>publish</wp:status>
    <wp:post_type>post</wp:post_type>
</item>
```

The related author in the `<dc:creator>` tag above defines the relationship between the blog post and the `<wp:author>` tag. The author data we need to construct the blog post and Author record in Wagtail is stored in a separate XML tag see below.

Sample XML tags for an author:

```xml
<wp:author>
  <wp:author_id>3</wp:author_id>
  <wp:author_login>joe-blogs</wp:author_login>
  <wp:author_email>inbox@example.com</wp:author_email>
  <wp:author_display_name>Joe Blogs</wp:author_display_name>
  <wp:author_first_name>Joe</wp:author_first_name>
  <wp:author_last_name>Blogs</wp:author_last_name>
</wp:author>
```

The `<wp:author>` XML tag above is not included in the import by default because only `<item>` tags that represent a blog post or page to be created in Wagtail are imported.

### XML Tag Hook Configuration

Add a settings variable to your site settings file called `WORDPRESS_IMPORT_HOOKS_TAGS_TO_CACHE` and add a dict with a key of `wp:author` and values as follows:

```python
WORDPRESS_IMPORT_HOOKS_TAGS_TO_CACHE = {

>     # the key is the XML tag name

Reviewer: **Suggestion**: Can we emphasise that this is the tag name of the target object in the relationship?

    "wp:author": {
        # the XML tag which has the value to lookup in the tags cache
        "DATA_TAG": "dc:creator",
        # the dotted path to the function to call in your own Wagtail site
        "FUNCTION": "pages.import_hooks.author_processor",
    }
}
```

The key of the dict is the name for the tag/s to cache in the XML file which contains the related content.

`DATA_TAG` is the tag `<dc:creator></dc:creator>` value. *This is passed to your function.*

`FUNCTION` is the dotted path to the function that will be called.

### Sample function for processing the author

```python
# pages/import_hooks.py

from pages.models import Author  # a snippet model for authors

> 

Reviewer: **Suggestion**: PEP-8 pedantry.

```suggestion

```
